modules/sql-engine/src/main/java/org/apache/ignite/internal/sql/engine/exec/ExchangeServiceImpl.java

```diff
@@ -162,6 +172,10 @@ private void onMessage(String nodeName, QueryBatchRequestMessage msg) {
     }
 
     private void onMessage(String nodeName, QueryBatchMessage msg) {
+        if (!nodeName.equals(localNodeName)) {
+            clock.update(msg.timestamp());
+        }
```

Comment:
I think this should be done at lower-level, e.g. somewhere in Messaging sub-system.
Shortly, a HLC (or similar) should register a hook in MessageService for that purpose, instead of shifting a responsibility to the other components. 

Let's add a TODO and create a ticket to fix this.

Reply:
Ticket created https://issues.apache.org/jira/browse/IGNITE-21709